Delayed sampling, testing framework, and verification of the concept paper results.

vsenderov opened the pull request:
# Updated description of the PR

1. Support for manual delayed sampling.
    a. Several new distributions to accommodate delayed sampling.
        - two new files `delayed.cuh` and `delayed.cu` that have special distributions dedicated to delayed sampling;
        - some extensions to the regular distributions: `lomax`, `negativeBinomial`, `chiSquared` and `student_t`; for `chiSquared` a specialized implementation for the GPU and CPU was attempted, but, in the end, only left the universal code; attempt can be supplied if needed.
     b. Some math functions: `lchoose`, `log`, `logBeta`.

2. Existing phylogenetic models simplified to contain only a simple CRBD implementation and the generic stack implementation of Joey from before. The advanced phylogenetics models will be organized separately.

3. Unit testing framework. To run the tests you need to do ./unit_tests.sh from RootPPL base. You can play with the test parameters inside the script.

4. NaN weight workaround by Joey.

david-broman commented on rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu:
```diff
@@ -0,0 +1,339 @@
+/*
+ * File clads2.cu defines the ClaDS2 model
```
There are several files under a folder `junk`. Should they really be committed? If so, what is the purpose of the `junk` folder?

vsenderov replied:
I briefly touched upon this in my PR description. This is one of the reasons why my initial idea was to discuss with Joey, before making the final PR. In a nutshell this is either older reference code, that I felt uneasy to delete, or some promising attempts, that have not been fully tested. They should go in a different branch, I suppose.

vsenderov replied:
Comments on how to handle it, appreciated.

JoeyOhman replied:
Many of these files are my old files. There are only two of these I think should stay:

1. `crbd/junk/condbd.cu`
2. `crbd/junk/stack.cuh`

These could turn out to be useful for reference when Daniel want to generate RootPPL code. However, they should eventually be removed as well. A directory for them, for now could be `crbd/generic_stack/`? Feel free to remove the rest of my old phylogenetics reference files that are no longer used. 

I do not know what to do with your files in the `junk` directories.

vsenderov replied:
We refactored those.

david-broman replied:
We need to rename the folder and add a good description of what these files are. I do not like to add folders with 'junk' in the repo.

JoeyOhman commented:
- Many new files are named with `kebab-case`. But the current file naming convention in RootPPL is `snake_case`. (Note that directories are named with `kebab-case` though)

- Is there a dependency on `R` for the unit tests? Perhaps it does not matter, but I had trouble executing them.

vsenderov commented:
@david-broman 

I cleaned up my PR, removed the junk folders, changed the filename case as it should be as per yours and Joey's suggestions.

The number of phylogenetic models has also been decreased. Some of the working ideas have been moved away. Now we have only tested models.

- CRBD with immediate sampling
- CRBD with delayed sampling (survivorship bias not implemented yet)
- ClaDS2 with immediate sampling (not saving intermediate states)
- ClaDS2 (factorized) with immediate sampling (not saving intermediate states)
- ClaDS2 (factorized) with delayed sampling (saving intermediate states)

I also removed the verification of the concept paper. This should go in a separate repo as it is not an integral part of RootPPL.

dlunde commented on rootppl/dists/dists_cpu.cuh:
```diff
@@ -60,4 +60,21 @@ unsigned int poisson(double lambda) {
+//   #else
+//   return dist(gen);
+//   #endif
+// }
```
Should this commented out code be removed?

vsenderov replied:
done

dlunde commented on rootppl/dists/dists_gpu.cuh:
```diff
@@ -84,4 +84,8 @@ __device__ floating_t gamma(curandState* randState, floating_t k, floating_t the
+
+
+
+#endif
```
Restore original white space here (no actual change to this file)

vsenderov replied:
done

dlunde commented on rootppl/do_birds.sh:
```diff
@@ -0,0 +1,19 @@
+    ./program $NUMPART $NUMRUN
+    cp log_norm_const.txt $RESULTDIR/$model.logz
+    mv results/EXP-* $RESULTDIR/
+done
```
Should this file really be in the top directory? To me, it feels like this should go under `models`. Or, maybe we should even consider moving the phylogenetic stuff to it's own repository.

vsenderov replied:
this is removed now

dlunde commented on rootppl/log_norm_const.txt:
```diff
@@ -0,0 +1 @@
+[-139.844]
```
This file should probably be deleted?

vsenderov replied:
done

dlunde reviewed (commented):
I think this can now be merged after some minor changes.

vsenderov commented:
> * Many new files are named with `kebab-case`. But the current file naming convention in RootPPL is `snake_case`. (Note that directories are named with `kebab-case` though)
> * Is there a dependency on `R` for the unit tests? Perhaps it does not matter, but I had trouble executing them.

I think I mostly deleted kebab-case-named-files. However, still some camelCaseFiles remain in the tests directory. Let me know if still problem.

vsenderov commented:
I addressed all issues in the comments.

In addition to that I back-ported the new CRBD version with "Jan's walk" and changed the default output of `smc.cu` to be the normalizing const without the verbosity (just a number).

dlunde commented:
I believe this can be merged now. @vsenderov I'll present it on the Miking meeting today

david-broman commented:
The main description should probably be updated. It still says that there are junk folders.

Also, how are the test models organized now? Are there duplications with the benchmark?

vsenderov commented:
# Updated description of the PR

1. Support for manual delayed sampling.
    a. Several new distributions to accommodate delayed sampling.
        - two new files `delayed.cuh` and `delayed.cu` that have special distributions dedicated to delayed sampling;
        - some extensions to the regular distributions: `lomax`, `negativeBinomial`, `chiSquared` and `student_t`; for `chiSquared` a specialized implementation for the GPU and CPU was attempted, but, in the end, only left the universal code; attempt can be supplied if needed.
     b. Some math functions: `lchoose`, `log`, `logBeta`.

2. Existing phylogenetic models simplified to contain only a simple CRBD implementation and the generic stack implementation of Joey from before. The advanced phylogenetics models will be organized a git submodule of `miking-benchmarks`.

3. Unit testing framework. To run the tests you need to do ./unit_tests.sh from RootPPL base. You can play with the test parameters inside the script.

4. NaN weight workaround by Joey.

dlunde commented:
@vsenderov Can you replace you original message with the above message instead? Easier to find that way, if we need to revisit this PR in the future.

vsenderov commented:
> @vsenderov Can you replace you original message with the above message instead? Easier to find that way, if we need to revisit this PR in the future.

Like so?

dlunde commented:
Yes, looks good!
